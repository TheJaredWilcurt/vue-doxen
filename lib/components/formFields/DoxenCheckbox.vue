<template>
  <FormFieldsetWrapper
    :modelValue="modelValue"
    :styleTokens="styleTokens"
  >
    <FormFieldLabel
      :disabled="disabled"
      :errorMessage="errorMessage"
      :label="label"
      :required="required"
      :styleTokens="styleTokens"
    />
    <div v-bind="applyStyleTokens({ formFieldCheckboxContainer: true })">
      <input
        v-bind="{
          ...$attrs,
          ...applyStyleTokens({
            formFieldCheckbox: true,
            formFieldCheckboxError: errorMessage
          })
        }"
        :id="idFor"
        :aria-invalid="!!errorMessage"
        :aria-required="!!required"
        :checked="modelValue"
        :data-test="idFor"
        :data-value="dataValue(modelValue)"
        :disabled="disabled"
        :name="idFor"
        :required="required"
        type="checkbox"
        @input="updateValue"
      />
      <label
        v-bind="applyStyleTokens({
          formFieldCheckboxNameLabel: true,
          formFieldCheckboxNameDisabled: disabled,
          formFieldCheckboxNameError: errorMessage
        })"
        :for="idFor"
      >
        {{ name }}
        <span
          v-if="!label && required"
          v-text="'*'"
          v-bind="applyStyleTokens({ formFieldRequired: true })"
        ></span>
      </label>
    </div>
    <FormFieldFooter
      :errorMessage="errorMessage"
      :innerHTML="$attrs.innerHTML"
      :message="message"
      :styleTokens="styleTokens"
    />
    <slot></slot>
  </FormFieldsetWrapper>
</template>

<script>
import {
  createIdFor,
  createImportStatement,
  generateRandomId
} from '@/helpers/componentHelpers.js';
import {
  createDisabledProp,
  createErrorMessageProp,
  createMessageProp,
  createModelValueProp,
  label,
  required,
  styleTokens
} from '@/helpers/props.js';
import { dataValue } from '@/helpers/snapshotHelpers.js';

import applyStyleTokens from '@/mixins/applyStyleTokensMixin.js';

import FormFieldFooter from '@/components/formFields/FormFieldFooter.vue';
import FormFieldLabel from '@/components/formFields/FormFieldLabel.vue';
import FormFieldsetWrapper from '@/components/formFields/FormFieldsetWrapper.vue';

const COMPONENT_NAME = 'DoxenCheckbox';
const disabled = createDisabledProp('checkbox');
const errorMessage = createErrorMessageProp('checkbox');
const message = createMessageProp('checkbox');
const modelValue = createModelValueProp(Boolean);

export default {
  ...createImportStatement(COMPONENT_NAME),
  name: COMPONENT_NAME,
  description: 'A checkbox form field component.',
  components: {
    FormFieldFooter,
    FormFieldLabel,
    FormFieldsetWrapper
  },
  mixins: [
    applyStyleTokens
  ],
  inheritAttrs: false,
  emits: ['update:model-value'],
  emitsToDemo: {
    'update:model-value': {
      value: 'A boolean matching if the checkbox is currently checked.'
    }
  },
  props: {
    disabled,
    errorMessage,
    label,
    message,
    modelValue,
    required,
    styleTokens,
    name: {
      type: String,
      default: undefined,
      description: 'A clickable label placed next to the checkbox.'
    }
  },
  methods: {
    dataValue,
    updateValue: function ($event) {
      this.$emit('update:model-value', $event.target.checked);
    }
  },
  computed: {
    uniqueId: function () {
      return generateRandomId();
    },
    idFor: function () {
      return createIdFor({
        label: this.label,
        value: 'boolean',
        name: this.name,
        uniqueId: this.uniqueId
      });
    }
  }
};
</script>
