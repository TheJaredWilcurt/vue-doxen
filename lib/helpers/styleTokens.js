/**
 * @file Style tokens allow a lot of flexibility in how the Vue-Doxen components can be styled.
 *
 * Style tokens solve the following scenarios:
 *
 * Provide optional basic built in CSS
 * Allow overriding basic built-in classes
 * The CSS is externalized and can be skipped if not needed
 * You can remove all classes from all elements if not needed
 * Allows customizing the class names to whatever you want
 * Allows having many classes applied on a single element (Tailwind)
 * Can use existing CSS class names (Bootstrap)
 * Can use existing CSS component classes (Vuetify)
 */

/**
 * A predictable map of tokens to class names. These classes are not used anywhere, they exist to make styling easier for those writing their own CSs.
 *
 * @type {Object}
 */
export const styleTokensBuiltIn = {
  button: 'doxen-button',
  buttonNotSelected: 'doxen-button-not-selected',
  buttonSelected: 'doxen-button-selected',
  codeBox: 'doxen-code-box',
  codeSwapper: 'doxen-code-swapper',
  codeSwapperFileName: 'doxen-code-swapper-file-name',
  componentDemo: 'doxen-component-demo',
  componentDemoContainer: 'doxen-component-demo-container',
  componentDemoH3: 'doxen-component-demo-h3',
  componentDemoHr: 'doxen-component-demo-hr',
  demoDescription: 'doxen-demo-description',
  demoHeader: 'doxen-demo-header',
  demoHeaderTitle: 'doxen-demo-header-title',
  emitsDocumentationCode: 'doxen-emits-documentation-code',
  emitsDocumentationLi: 'doxen-emits-documentation-li',
  emitsDocumentationSpan: 'doxen-emits-documentation-span',
  emitsDocumentationStrong: 'doxen-emits-documentation-strong',
  emitsDocumentationUl: 'doxen-emits-documentation-ul',
  emitLogClearButton: 'doxen-emit-log-clear-button',
  emitLogHeader: 'doxen-emit-log-header',
  emitLogTable: 'doxen-emit-log-table',
  emitLogTableContainer: 'doxen-emit-log-table-container',
  emitLogTbody: 'doxen-emit-log-tbody',
  emitLogTd: 'doxen-emit-log-td',
  emitLogTh: 'doxen-emit-log-th',
  emitLogThead: 'doxen-emit-log-thead',
  emitLogTr: 'doxen-emit-log-tr',
  formFieldCheckbox: 'doxen-form-field-checkbox',
  formFieldCheckboxContainer: 'doxen-form-field-checkbox-container',
  formFieldCheckboxError: 'doxen-form-field-checkbox-error',
  formFieldCheckboxNameDisabled: 'doxen-form-field-checkbox-name-disabled',
  formFieldCheckboxNameError: 'doxen-form-field-checkbox-name-error',
  formFieldCheckboxNameLabel: 'doxen-form-field-checkbox-name-label',
  formFieldDropdown: 'doxen-form-field-dropdown',
  formFieldDropdownContainer: 'doxen-form-field-dropdown-container',
  formFieldDropdownError: 'doxen-form-field-dropdown-error',
  formFieldDropdownOption: 'doxen-form-field-dropdown-option',
  formFieldError: 'doxen-form-field-error',
  formFieldFieldset: 'doxen-form-field-fieldset',
  formFieldLabel: 'doxen-form-field-label',
  formFieldLegend: 'doxen-form-field-legend',
  formFieldLegendDisabled: 'doxen-form-field-legend-disabled',
  formFieldLegendError: 'doxen-form-field-legend-error',
  formFieldMessage: 'doxen-form-field-message',
  formFieldRadioDial: 'doxen-form-field-radio-dial',
  formFieldRadioDialContainer: 'doxen-form-field-radio-dial-container',
  formFieldRadioDialNameDisabled: 'doxen-form-field-radio-dial-name-disabled',
  formFieldRadioDialNameError: 'doxen-form-field-radio-dial-name-error',
  formFieldRadioDialNameLabel: 'doxen-form-field-radio-dial-name-label',
  formFieldRadioDialsContainer: 'doxen-form-field-radio-dials-container',
  formFieldRequired: 'doxen-form-field-required',
  formFieldTextarea: 'doxen-form-field-textarea',
  formFieldTextareaCode: 'doxen-form-field-textarea-code',
  formFieldTextareaContainer: 'doxen-form-field-textarea-container',
  formFieldTextareaError: 'doxen-form-field-textarea-error',
  formFieldTextField: 'doxen-form-field-text-field',
  formFieldTextFieldContainer: 'doxen-form-field-text-field-container',
  formFieldTextFieldError: 'doxen-form-field-text-field-error',
  propsDocumentationCode: 'doxen-props-documentation-code',
  propsDocumentationEm: 'doxen-props-documentation-em',
  propsDocumentationLi: 'doxen-props-documentation-li',
  propsDocumentationSpan: 'doxen-props-documentation-span',
  propsDocumentationStrong: 'doxen-props-documentation-strong',
  propsDocumentationUl: 'doxen-props-documentation-ul',
  propsPlaygroundForm: 'doxen-props-playground-form',
  sidebarButton: 'doxen-sidebar-button',
  sidebarButtonNotSelected: 'doxen-sidebar-button-not-selected',
  sidebarButtonSelected: 'doxen-sidebar-button-selected',
  sidebarContainer: 'doxen-sidebar-container',
  tabButton: 'doxen-tab-button',
  tabButtonNotSelected: 'doxen-tab-button-not-selected',
  tabButtonSelected: 'doxen-tab-button-selected',
  tabsButtonContainer: 'doxen-tabs-button-container',
  tabsHeader: 'doxen-tabs-header',
  vueDoxen: 'doxen-vue-doxen'
};

/**
 * All the style token keys without any class names applied to them.
 *
 * @type {Object}
 */
export const styleTokensEmpty = {};
for (const key in styleTokensBuiltIn) {
  styleTokensEmpty[key] = '';
}

/**
 * Style tokens using the class names provided by Bootstrap 5.
 *
 * @type {Object}
 */
export const styleTokensBootstrap5 = {
  // Container
  formFieldFieldset: 'px-2 pb-2 m-1 flex-grow-1',
  // Errors
  formFieldCheckboxError: 'text-danger',
  formFieldCheckboxNameError: 'text-danger',
  formFieldDropdownError: 'text-danger',
  formFieldError: 'form-text text-danger',
  formFieldLegendError: 'text-danger',
  formFieldRadioDialNameError: 'text-danger',
  formFieldTextareaError: 'text-danger',
  formFieldTextFieldError: 'text-danger',
  // Labels
  formFieldLegend: 'form-label form-control-sm mb-0 px-0',
  // Inputs
  formFieldTextarea: 'form-control',
  formFieldTextField: 'form-control',
  // Checkbox
  formFieldCheckbox: 'form-check-input',
  formFieldCheckboxContainer: 'form-check',
  formFieldCheckboxNameDisabled: '',
  formFieldCheckboxNameLabel: 'form-check-label',
  // Radio Dials
  formFieldRadioDial: 'form-check-input',
  formFieldRadioDialNameLabel: 'form-check-label form-label form-control-sm',
  formFieldRadioDialContainer: 'form-check',
  // Select
  formFieldDropdown: 'form-select form-select-sm',
  // Message
  formFieldMessage: 'form-text',
  // Buttons
  button: 'btn',
  buttonSelected: 'btn-primary',
  buttonNotSelected: 'btn-outline-primary',
  // Code
  codeSwapperFileName: 'float-end',
  formFieldTextareaCode: 'font-monospace',
  tabsButtonContainer: 'nav nav-tabs nav-item mt-3 mb-0',
  tabButton: 'nav-link me-1',
  tabButtonSelected: 'active border-bottom-0 mb-0',
  // Demo
  propsPlaygroundForm: 'd-flex flex-wrap',
  // Emit log
  emitLogHeader: 'd-flex',
  emitLogClearButton: 'btn btn-primary',
  emitLogTable: 'table table-striped',
  // Sidebar
  sidebarContainer: 'btn-group-vertical mx-3 justify-content-start',
  sidebarButton: 'btn flex-grow-0',
  sidebarButtonNotSelected: 'btn-outline-primary',
  sidebarButtonSelected: 'btn-primary flex-shrink-1'
};
