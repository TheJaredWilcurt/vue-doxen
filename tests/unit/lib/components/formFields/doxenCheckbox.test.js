import { styleTokensBuiltIn } from '@/helpers/styleTokens.js';

import DoxenCheckbox from '@/components/formFields/DoxenCheckbox.vue';

import testHelpers from '@@/unit/testHelpers.js';

describe('DoxenCheckbox.vue', () => {
    const styleTokens = styleTokensBuiltIn;
    const requiredProps = {};

    const setupWrapper = async (props) => {
        const options = {
            props: {
                ...requiredProps,
                ...props
            },
            slots: {
                default: 'Slot Text'
            }
        };
        const wrapper = await testHelpers.mount(DoxenCheckbox, options);
        return wrapper;
    };

    test('Renders correctly', async () => {
        const wrapper = await setupWrapper();

        expect(wrapper)
            .toMatchSnapshot();
    });

    test('Style tokens', async () => {
        const wrapper = await setupWrapper({ styleTokens });

        expect(wrapper)
            .toMatchSnapshot();
    });

    test('Emit click', async () => {
        const wrapper = await setupWrapper();

        await wrapper.find('input').trigger('click');

        expect(wrapper.emitted().click[0].length)
            .toEqual(1);
    });

    test('idFor should equal input name', async () => {
        const wrapper = await setupWrapper();

        const checkboxInput = wrapper.find('input[type="checkbox"]');
        const labelElement = wrapper.find('label');
    
        // Ensure both elements exist
        expect(checkboxInput.exists()).toBe(true);
        expect(labelElement.exists()).toBe(true);
    
        const checkboxName = checkboxInput.attributes('name');
        const labelFor = labelElement.attributes('for');
    
        // Compare the attributes
        expect(checkboxName).toEqual(labelFor);
    })
})