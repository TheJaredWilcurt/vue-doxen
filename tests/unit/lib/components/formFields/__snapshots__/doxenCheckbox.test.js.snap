// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`DoxenCheckbox.vue > Renders correctly 1`] = `
<fieldset
  class=""
  data-applied-style-tokens="formFieldFieldset"
  data-style-tokens="formFieldFieldset"
  data-value="undefined"
>
  <!-- v-if -->
  <div
    class=""
    data-applied-style-tokens="formFieldCheckboxContainer"
    data-style-tokens="formFieldCheckboxContainer"
  >
    <input
      aria-invalid="false"
      aria-required="false"
      class=""
      data-applied-style-tokens="formFieldCheckbox"
      data-style-tokens="formFieldCheckbox formFieldCheckboxError"
      data-value="undefined"
      id="VueDoxen_boolean_v-0"
      name="VueDoxen_boolean_v-0"
      type="checkbox"
      value="'on'"
    />
    <label
      class=""
      data-applied-style-tokens="formFieldCheckboxNameLabel"
      data-style-tokens="formFieldCheckboxNameLabel formFieldCheckboxNameDisabled formFieldCheckboxNameError"
      for="VueDoxen_boolean_v-0"
    >
      <!-- v-if -->
    </label>
  </div>
  <!-- v-if -->
  <!-- v-if -->
  <!-- v-if -->
  Slot Text
</fieldset>
`;

exports[`DoxenCheckbox.vue > Style tokens 1`] = `
<fieldset
  class="doxen-form-field-fieldset"
  data-applied-style-tokens="formFieldFieldset"
  data-style-tokens="formFieldFieldset"
  data-value="undefined"
>
  <!-- v-if -->
  <div
    class="doxen-form-field-checkbox-container"
    data-applied-style-tokens="formFieldCheckboxContainer"
    data-style-tokens="formFieldCheckboxContainer"
  >
    <input
      aria-invalid="false"
      aria-required="false"
      class="doxen-form-field-checkbox"
      data-applied-style-tokens="formFieldCheckbox"
      data-style-tokens="formFieldCheckbox formFieldCheckboxError"
      data-value="undefined"
      id="VueDoxen_boolean_v-0"
      name="VueDoxen_boolean_v-0"
      type="checkbox"
      value="'on'"
    />
    <label
      class="doxen-form-field-checkbox-name-label"
      data-applied-style-tokens="formFieldCheckboxNameLabel"
      data-style-tokens="formFieldCheckboxNameLabel formFieldCheckboxNameDisabled formFieldCheckboxNameError"
      for="VueDoxen_boolean_v-0"
    >
      <!-- v-if -->
    </label>
  </div>
  <!-- v-if -->
  <!-- v-if -->
  <!-- v-if -->
  Slot Text
</fieldset>
`;
